File: frontend/src/lib/usernameColor.ts
/**
 * Stable HSL color for a chat participant. Prefer `senderId` when present so
 * renames do not change the color; otherwise hash the display name.
 */
export function usernameColorFromUser(
  senderId: number | undefined,
  displayName: string
): string {
  const seed =
    senderId != null ? `id:${senderId}` : `name:${displayName.toLowerCase().trim()}`;
  return hashSeedToHsl(seed);
}

function hashSeedToHsl(seed: string): string {
  let h = 2166136261;
  for (let i = 0; i < seed.length; i++) {
    h ^= seed.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  const hue = Math.abs(h) % 360;
  return `hsl(${hue} 62% 70%)`;
}

File: android/app/src/main/java/com/chatsib/app/core/UsernameColor.kt
package com.chatsib.app.core

import androidx.compose.ui.graphics.Color

/**
 * Stable HSL color for a chat participant. Prefer [senderId] when present so
 * renames do not change the color; otherwise hash the display name.
 * Port of `frontend/src/lib/usernameColor.ts`.
 */
fun usernameColorFromUser(senderId: Int?, displayName: String): Color {
    val seed = if (senderId != null) {
        "id:$senderId"
    } else {
        "name:${displayName.lowercase().trim()}"
    }
    return hashSeedToHsl(seed)
}

private fun hashSeedToHsl(seed: String): Color {
    var h = 2166136261L
    for (ch in seed) {
        h = h xor ch.code.toLong()
        h = (h * 16777619L) and 0xFFFFFFFFL
    }
    val hue = kotlin.math.abs(h.toInt()) % 360
    return Color.hsl(hue.toFloat(), 0.62f, 0.70f)
}
